src/crab/type_domain.cpp

```diff
@@ -0,0 +1,714 @@
+                auto ptr_with_off2 = std::get<crab::ptr_with_off_t>(p2);
+                return (ptr_with_off1.r == ptr_with_off2.r && ptr_with_off1.offset == ptr_with_off2.offset);
+            }
+        }
+    };
+}
+
+
+void print_ptr_type(const ptr_t& p) {
+    if (std::holds_alternative<ptr_with_off_t>(p)) {
+        auto t = std::get<ptr_with_off_t>(p);
+        std::cout << t;
+    }
+    else {
+        auto t = std::get<ptr_no_off_t>(p);
+        std::cout << t;
+    }
+}
+
+void print_type(register_t r, const ptr_t& p) {
+    std::cout << "r" << static_cast<unsigned int>(r) << " : ";
+    print_ptr_type(p);
+}
+
+void print_annotated(Mem const& b, const ptr_t& p, std::ostream& os_) {
+    if (b.is_load) {
+        os_ << "  ";
+        print_type(std::get<Reg>(b.value).v, p);
+        os_ << " = ";
+    }
+    std::string sign = b.access.offset < 0 ? " - " : " + ";
+    int offset = std::abs(b.access.offset);
+    os_ << "*(" << size(b.access.width) << " *)";
+    os_ << "(" << b.access.basereg << sign << offset << ")\n";
+}
+
+void print_annotated(Call const& call, const ptr_t& p, std::ostream& os_) {
+    os_ << "  ";
+    print_type(0, p);
+    os_ << " = " << call.name << ":" << call.func << "(...)\n";
+}
+
+namespace crab {
+
+inline std::string get_reg_ptr(const region& r) {
+    switch (r) {
+        case region::T_CTX:
+            return "ctx_p";
+        case region::T_STACK:
+            return "stack_p";
+        case region::T_PACKET:
+            return "packet_p";
+        default:
+            return "shared_p";
+    }
+}
+
+inline std::ostream& operator<<(std::ostream& o, const region& t) {
+    o << static_cast<std::underlying_type<region>::type>(t);
+    return o;
+}
+
+bool operator==(const ptr_with_off_t& p1, const ptr_with_off_t& p2) {
+    return (p1.r == p2.r && p1.offset == p2.offset);
+}
+
+bool operator!=(const ptr_with_off_t& p1, const ptr_with_off_t& p2) {
+    return !(p1 == p2);
+}
+
+std::ostream& operator<<(std::ostream& o, const ptr_with_off_t& p) {
+    o << get_reg_ptr(p.r) << "<" << p.offset << ">";
+    return o;
+}
+
+bool operator==(const ptr_no_off_t& p1, const ptr_no_off_t& p2) {
+    return (p1.r == p2.r);
+}
+
+bool operator!=(const ptr_no_off_t& p1, const ptr_no_off_t& p2) {
+    return !(p1 == p2);
+}
+
+std::ostream& operator<<(std::ostream& o, const ptr_no_off_t& p) {
+    return o << get_reg_ptr(p.r);
+}
+
+std::ostream& operator<<(std::ostream& o, const reg_with_loc_t& reg) {
+    o << "r" << static_cast<unsigned int>(reg.r) << "@" << reg.loc->second << " in " << reg.loc->first << " ";
+    return o;
+}
+
+bool reg_with_loc_t::operator==(const reg_with_loc_t& other) const {
+    return (r == other.r && loc == other.loc);
+}
+
+std::size_t reg_with_loc_t::hash() const {
+    // Similar to boost::hash_combine
+    using std::hash;
+
+    std::size_t seed = hash<register_t>()(r);
+    seed ^= hash<int>()(loc->first.from) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
```

Comment:
Isn't `loc->first` a basic block label? Label is already probably a string. What is `from` and `to`?

Reply:
No, label is not a string, and jump source (`from`) and jump target (`to`) are the only data that uniquely identify the label. Hence, I use `from` and `to` for hash function. I avoided modifying the `label_t` class as it is being used extensively in other parts of prevail.